### PR TITLE
[Merged by Bors] - feat(List): Miscellaneous lemmas

### PR DESCRIPTION
**Pull request by YaelDillies**
From LeanCamCombi

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Perm.lean [hunk 1]
```diff
@@ -245,14 +245,9 @@ attribute [simp] nil_subperm
...
 theorem subperm_nil : List.Subperm l [] ↔ l = [] :=
-  match l with
-  | [] => by simp
-  | head :: tail => by
-    simp only [iff_false]
-    intro h
-    have := h.length_le
-    simp only [List.length_cons, List.length_nil, Nat.succ_ne_zero, ← Nat.not_lt, Nat.zero_lt_succ,
-      not_true_eq_false] at this
+  ⟨fun h ↦ length_eq_zero.1 $ Nat.le_zero.1 h.length_le, by rintro rfl; rfl⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Perm.lean:248]:
```suggestion
  ⟨fun h ↦ length_eq_zero.1 <| Nat.le_zero.1 h.length_le, by rintro rfl; rfl⟩
```

File: Mathlib/Data/List/Basic.lean [hunk 2]
```diff
@@ -671,6 +671,20 @@ theorem getLast_replicate_succ (m : ℕ) (a : α) :
   exact getLast_append_singleton _
 #align list.last_replicate_succ List.getLast_replicate_succ
 
+lemma getLast_filter {p : α → Bool} :
+    ∀ (l : List α) (hlp : l.filter p ≠ []), p (l.getLast (hlp <| ·.symm ▸ rfl)) = true →
+      (l.filter p).getLast hlp = l.getLast (hlp <| ·.symm ▸ rfl)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Basic.lean:676]: Can you add a doc comment? It takes me altogether too long to parse the statement

File: Mathlib/Data/List/Basic.lean [hunk 3]
```diff
@@ -671,6 +671,20 @@ theorem getLast_replicate_succ (m : ℕ) (a : α) :
...
+  | [a], h, h' => by rw [List.getLast_singleton'] at h'; simp [List.filter_cons, h']
+  | a :: b :: as, h, h' => by
+    rw [List.getLast_cons_cons] at h' ⊢
+    simp only [List.filter_cons (x := a)] at h ⊢
+    obtain ha | ha := Bool.eq_false_or_eq_true (p a)
+    · simp only [ha, ite_true]
+      rw [getLast_cons, getLast_filter (b :: as) _ h']
+      exact ne_nil_of_mem $ mem_filter.2 ⟨getLast_mem _, h'⟩
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Basic.lean:684]:
```suggestion
      exact ne_nil_of_mem <| mem_filter.2 ⟨getLast_mem _, h'⟩
```

**Review by Ruben-VandeVelde — APPROVED** [review]

**Review by Ruben-VandeVelde — APPROVED** [review]
Thanks!

maintainer merge

**Comment by riccardobrasca** [comment]
Thanks!

bors merge

**Pull request closed without merging**